### PR TITLE
Gandalf scraper with tests

### PR DESCRIPTION
**Pull request by patkkka**

### CODE REVIEW

File: gandalfscraper/src/main/java/com/blackbook/gandalfscraper/scraper/BookPage.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+                .stream()
+                .map(Element::text)
+                .collect(Collectors.toList());
+        return String.join(", ", authorsList);
+    }
+
+    String extractBookGenre() {
+        int mainGenreIndex = 1;
+        Elements elements = bookDoc.select(".product_categories > a");
+        String genre = "nieznany";
```
Comment (piotr921) [inline gandalfscraper/src/main/java/com/blackbook/gandalfscraper/scraper/BookPage.java:53]: Change to "Unknown"

File: gandalfscraper/src/main/java/com/blackbook/gandalfscraper/GandalfScraperApp.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ * @author "Patrycja Zaremba"
+ */
+@SpringBootApplication
+@EnableAsync
+public class GandalfScraperApp {
+
+    @Bean
+    public BotService scrapperService(){
+        return new CrawlerScraperService(new Scraper(new JsoupWebConnector(), new LastPageChecker()));
+    }
```
Comment (Devas) [inline gandalfscraper/src/main/java/com/blackbook/gandalfscraper/GandalfScraperApp.java:23]: Enter XD

File: gandalfscraper/src/main/java/com/blackbook/gandalfscraper/controller/ScraperController.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+/**
+ * @author "Patrycja Zaremba"
+ */
+
+@Slf4j
+@RestController
+@RequestMapping(value = "/api/gandalf-scraper")
+public class ScraperController {
+
+    @Autowired
```
Comment (Devas) [inline gandalfscraper/src/main/java/com/blackbook/gandalfscraper/controller/ScraperController.java:21]: 👎 Don't autowire by field, use setter or constructor

File: gandalfscraper/src/main/java/com/blackbook/gandalfscraper/scraper/BookDiscountDataCreator.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        String title = bookDoc.extractBookTitle();
+        String subtitle = bookDoc.extractBookSubtitle();
+        String authors = bookDoc.extractBookAuthors();
+        String genre = bookDoc.extractBookGenre();
+        Double price = bookDoc.extractBookPrice();
+        String promoDetails = bookDoc.extractBookPromoDetails();
+        String coverUrl = bookDoc.extractBookCoverUrl();
+        String publisher = bookDoc.extractPublisher();
+
+        BookDiscountData bookDiscountData = BookDiscountData.builder()
```
Comment (Devas) [inline gandalfscraper/src/main/java/com/blackbook/gandalfscraper/scraper/BookDiscountDataCreator.java:24]: 👍

**Review by piotr921 — APPROVED** [review]

File: gandalfscraper/src/main/java/com/blackbook/gandalfscraper/scraper/Scraper.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+/**
+ * @author "Patrycja Zaremba"
+ */
+@Component
+public class Scraper implements ICrawler {
+    public static final int BOOKSTORE_ID = 4;
+
+    public static final String GANDALF_BOOKSTORE_URL = "http://www.gandalf.com.pl";
+    private static final String GANDALF_DISCOUNT_URL = GANDALF_BOOKSTORE_URL + "/k/okazje-do-60/";
```
Comment (Devas) [inline gandalfscraper/src/main/java/com/blackbook/gandalfscraper/scraper/Scraper.java:26]: 👍

File: gandalfscraper/src/test/java/com/blackbook/gandalfscraper/controller/ScraperControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    @MockBean
+    BotService botService;
+
+    @LocalServerPort
+    private int port;
+
+    @Test
+    public void shouldPostMatrasScraper() {
+        String matrasEndpointUrl = "/api/gandalf-scraper";
+        Mockito.doNothing().when(botService).saveResultsInDatabase();
```
Comment (Devas) [inline gandalfscraper/src/test/java/com/blackbook/gandalfscraper/controller/ScraperControllerTest.java:30]: 👍

File: gandalfscraper/src/test/java/com/blackbook/gandalfscraper/scraper/ScraperTest.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        when(lastPageChecker.extractLastPage(new Document(anyString()))).thenReturn(lastPageNo);
+    }
+
+    @Test
+    public void shouldScrapBooksFromGivenHTML(){
+        //given
+        int booksOnPage = 2;
+        Scraper scraper = new Scraper(webConnector, lastPageChecker);
+        //when
+        scraper.start(new CrawlerActionListener() {
```
Comment (Devas) [inline gandalfscraper/src/test/java/com/blackbook/gandalfscraper/scraper/ScraperTest.java:51]: lambda?

**Review by Devas — APPROVED** [review]

**Review by mprtcz — APPROVED** [review]

**Pull request merged**